File: IOSViperSample/IOSViperSample/chapters/ChapterListPresenterImpl.swift
//
//  ChapterListPresenterImpl.swift
//  TestTableView
//
//  Created by Ivan Phytsyk on 9/25/18.
//  Copyright © 2018 Ivan Phytsyk. All rights reserved.
//

import Foundation

class ChapterListPresenterImpl: ChapterListPresenter {
    
    private let chapterListInteractor: GetChaptersInteractor
    private let view: ChapterListView
    private let router: ChapterListRouter
    
    init(_ chapterListInteractor: GetChaptersInteractor
        ,_ view: ChapterListView
        ,_ router: ChapterListRouter) {
        self.chapterListInteractor = chapterListInteractor
        self.view = view
        self.router = router
    }
    
    func start() {
        chapterListInteractor.execute({
            let viewModels = $0?.map({ (chapter) -> ChapterViewModel in
                ChapterViewModel(chapter.title, chapter.description, getIconForChapterType(chapter.type))
            })
            self.view.showChapters(viewModels)
        }) {
            self.view.showErrorMessage($0.localizedDescription)
        }
    }
    
    func stop() {
        chapterListInteractor.cancel()
    }
    
    func chapterItemClicked(_ chapter: ChapterViewModel) {
        router.openChapterDetails(chapter.title!)
    }
}

File: AndroidViperSample/app/src/main/java/fytsyk/com/vipersample/chapters/ChapterListPresenterImpl.kt
package fytsyk.com.vipersample.chapters

import fytsyk.com.vipersample.chapters.viper.*

class ChapterListPresenterImpl(private val chapterListInteractor: GetChaptersInteractor,
                               private val view: ChapterListView,
                               private val router: ChapterListRouter) : ChapterListPresenter {


    override fun start() {
        view.showProgress(true)
        chapterListInteractor.execute(null, { chapter ->
            val viewModels = chapter?.map {
                ChapterViewModel(it.title,
                        it.description, getIconForChapterType(it.type))
            }
            view.showChapters(viewModels)
        }) {
            view.showErrorMessage(it.message)
        }
    }

    override fun stop() {
        chapterListInteractor.cancel()
    }

    override fun chapterItemClicked(chapter: ChapterViewModel) {
        router.openChapterDetails(chapter)
    }
}